resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/submitsteps/hadoopsteps/HadoopKerberosKeytabResolverStep.scala

```diff
@@ -0,0 +1,125 @@
+  * you then retrieve the delegation token from the NameNode and store values in
+  * DelegationToken. Lastly, the class puts the data into a secret. All this is
+  * appended to the current HadoopSpec which in turn will append to the current
+  * DriverSpec.
+  */
+private[spark] class HadoopKerberosKeytabResolverStep(
+  submissionSparkConf: SparkConf,
+  maybePrincipal: Option[String],
+  maybeKeytab: Option[File],
+  maybeRenewerPrincipal: Option[String],
+  hadoopUGI: HadoopUGIUtil) extends HadoopConfigurationStep with Logging {
+    private var originalCredentials: Credentials = _
+    private var dfs : FileSystem = _
+    private var renewer: String = _
+    private var credentials: Credentials = _
+    private var tokens: Iterable[Token[_ <: TokenIdentifier]] = _
+
+    override def configureContainers(hadoopConfigSpec: HadoopConfigSpec): HadoopConfigSpec = {
+      val hadoopConf = SparkHadoopUtil.get.newConfiguration(submissionSparkConf)
+      if (!hadoopUGI.isSecurityEnabled) logDebug("Hadoop not configured with Kerberos")
+      val maybeJobUserUGI =
+        for {
+          principal <- maybePrincipal
+          keytab <- maybeKeytab
+        } yield {
+          // Not necessary with [Spark-16742]
+          // Reliant on [Spark-20328] for changing to YARN principal
+          submissionSparkConf.set("spark.yarn.principal", principal)
+          submissionSparkConf.set("spark.yarn.keytab", keytab.toURI.toString)
+          logDebug("Logged into KDC with keytab using Job User UGI")
+          hadoopUGI.loginUserFromKeytabAndReturnUGI(
+            principal,
+            keytab.toURI.toString)
+        }
+      // In the case that keytab is not specified we will read from Local Ticket Cache
+      val jobUserUGI = maybeJobUserUGI.getOrElse(hadoopUGI.getCurrentUser)
+      // It is necessary to run as jobUserUGI because logged in user != Current User
+      jobUserUGI.doAs(new PrivilegedExceptionAction[Void] {
+        override def run(): Void = {
+          originalCredentials = jobUserUGI.getCredentials
+          // TODO: This is not necessary with [Spark-20328] since we would be using
+          // Spark core providers to handle delegation token renewal
+          renewer = maybeRenewerPrincipal.getOrElse(jobUserUGI.getShortUserName)
+          credentials = new Credentials(originalCredentials)
+          hadoopUGI.dfsAddDelegationToken(hadoopConf, renewer, credentials)
+          tokens = credentials.getAllTokens.asScala
+          null
+        }})
+      if (tokens.isEmpty) logDebug("Did not obtain any Delegation Tokens")
+      val data = hadoopUGI.serialize(credentials)
+      val renewalInterval =
+        hadoopUGI.getTokenRenewalInterval(tokens, hadoopConf).getOrElse(Long.MaxValue)
+      val currentTime: Long = hadoopUGI.getCurrentTime
+      val initialTokenDataKeyName = s"$KERBEROS_SECRET_LABEL_PREFIX-$currentTime-$renewalInterval"
+      val uniqueSecretName = s"$HADOOP_KERBEROS_SECRET_NAME.$currentTime"
```

Comment:
Can we name this using $kubernetesResourceNamePrefix like the hadoop config map name so it's easier to tell which secret is for which Spark job?